### PR TITLE
Add explicit interfaces for Shapers

### PR DESCRIPTION
**Pull request by tuespetre**
Another breakout from #7543, this time isolating about 9 files. 

- Added an `IShaper` interface and changed casts/tests from `Shaper` to `IShaper`
- Added an `IOffsettableShaper` interface with an `OffsettableShaper` base class and changed casts/tests from `EntityShaper` to `IOffsettableShaper`

### CODE REVIEW

**Comment by tuespetre** [comment]
Hey, who got Appveyor to work? 😉

**Comment by maumar** [comment]
What is the reason for this change? It doesn't seem to provide much value.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs [hunk 1]
```diff
@@ -239,9 +239,9 @@ var useQueryComposition
                 Expression.Constant(shaper));
         }
 
-        private Shaper CreateShaper(Type elementType, IEntityType entityType, SelectExpression selectExpression)
+        private IShaper CreateShaper(Type elementType, IEntityType entityType, SelectExpression selectExpression)
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs]: Change to interface is problematic here. We won't be able to add functionality to it without introducing a breaking change. Abstract base class doesn't have this limitation.

**Comment by tuespetre** [comment]
You may be right, part of why I split these changes out. I thought I had more use for them with some other Shaper classes but that may not be valid anymore. I will double check and close this if that's the case

**Comment by tuespetre** [comment]
Sorry for wasting your time with this one! 🙀

**Pull request closed without merging**